docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,234 @@
+* Approver: @mortent
+
+## Why
+
+When deploying workloads across large fleets of clusters, it is often necessary
+to modify the workload configuration for a specific cluster. Additionally, those
+workloads may evolve over time with security or other patches that require
+updates. [Configuration as Data](06-config-as-data.md) in general and [Package
+Orchestration](07-package-orchestration.md) in particular can assist in this.
+However, they are still centered around manual, one-by-one hydration and
+configuration of a workload.
+
+This proposal introduces concepts and a set of resources for automating the
+creation and lifecycle management of package variants. These are designed to
+address several different dimensions of scalability:
+- Number of different workloads for a given cluster
+- Number of clusters across which those workloads are deployed
+- Complexity of the organizations deploying those workloads (NOTE: actually this
+  is more the `conditions` stuff, will probably move this out of here and create
+  a separate proposal for that)
+- Changes to those workloads over time
+
+## See Also
+- [Package Orchestration](07-package-orchestration.md)
+- [#3488](https://github.com/GoogleContainerTools/kpt/issues/3488)
+- [Managing Package
+   Revisions](https://docs.google.com/document/d/1EzUUDxLm5jlEG9d47AQOxA2W6HmSWVjL1zqyIFkqV1I/edit?usp=sharing)
+- [Porch UpstreamPolicy Resource
+  API](https://docs.google.com/document/d/1OxNon_1ri4YOqNtEQivBgeRzIPuX9sOyu-nYukjwN1Q/edit?usp=sharing&resourcekey=0-2nDYYH5Kw58IwCatA4uDQw)
+
+## Core Concepts
+
+For this solution, "workloads" are represented by packages. "Package" is a more
+general concept, being an arbitrary bundle of resources, and therefore is
+sufficient to solve the originally stated problem.
+
+The basic idea here is to introduce a `PackageVariant` resource that manages the
+derivation of variant of a package from the original source package, and to
+manage the evolution of that variant over time. This effectively automates the
+human-centered process for variant creation one might use with `kpt`:
+1. Clone and upstream package locally
+1. Make changes to the local package, including executing KRM functions
+1. Push the package to a new repository and tag it as a new version
+
+Similarly, `PackageVariant` can manage the process of updating a package when a
+new version of the upstream package is published. In the human-centered
+workflow, a user would use `kpt pkg update` to pull in changes to their
+derivative package. When using a `PackageVariant` resource, the change would be
+made to the upstream specification in the resource, and the controller would
+propose a new Draft package reflecting the outcome of `kpt pkg update`.
+
+By automating this process, we open up the possibility of performing systematic
+changes that tie back to our different dimensions of scalability. For example,
+we can use data about the specific variant we are creating to lookup additional
```

Comment:
I like thinking about this as a continuous reconciliation where the inputs are a parent package `P` and a context `C` and the output being the derived package `D`. And every time either the version of the upstream package or the context changes, we will get a new revision of the derived package that has the provided revision of the upstream package as upstream and has been customized according to the context.

Reply:
Updated to reflect this.